study/hello-caqti-eio/README.md

```diff
@@ -0,0 +1,53 @@
+# Hello, caqti-eio!
+
+Let's look at [eio](https://github.com/ocaml-multicore/eio)!
+
+Instead of promises, we will instead use a "direct style" concurrency model.
+
+> NOTE: OCaml v5 or greater is required.
+
+As always let's run our tests:
+
+```
+cd ./hello-caqti-eio
+PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune exec ./bin/main.exe
+PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune runtest --watch
+```
+
+## Test via the REPL
+
+
+```
+$ PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune utop
+```
+
+Let's bring the `Init` and `Exec` modules into scope:
+```ocaml
+# open Repo
+```
+
+Now let's query the database!
+
+Unlike previously, we have to conform to a callback style. See the [Exec](./lib/repo/exec.ml) module for more info.
```

Comment:
I think I actually made a little mistake in the first tutorials by trying to "hold on" to a connection (mainly for the REPL interactions)

I think I should probably update the tutorials so that I delegate every time to:

```
Caqti_***.with_connection
``` 

As I did here. This would make things more cohesive, overall. Do you agree ?

By declaring the `with_conn` helper function, I found it quite nice to play around with the repo functions in the REPL. Also, it seems more robust.

```
with_conn @@ Exec.add 1 2;;
```

Reply:
The style adopted in EIO's own `README.md` is to wrap inputs in `Eio_main.run @@ fun env ->` (and a nested similar where a switch is needed), so I guess that's our option if we don't want to bake `Eio_main.run` into `with_conn`.  It's somewhat verbose.  It could be abbreviated to something like `run @@ with_conn @@ Exec.add 1 2`, I think, where `run` would pass the environment and switch, since named arguments can be commuted over the unnamed one.

Reply:
> so I guess that's our option if we don't want to bake Eio_main.run into with_conn

If you look at my last commit https://github.com/paurkedal/caqti-study/pull/12/commits/0e0d6947670fa1afdfdbcc15dbe4856674ee6586#diff-2c6fb6befa514f878d5824655c21c0fe37cc95773e733f3f2c9e0bef5bc15a1eR25

I already took out `Eio_main.run` of `with_conn`

I played with API to make it fit the EIO style looking at its README, it could look something like this:

```ocaml
let main ~(env : Eio_unix.Stdenv.base) =
  let ( let* ) = Result.bind in
  let program : (data, 'err) result =
    Init.with_conn ~stdenv:(env :> Caqti_eio.stdenv) @@ fun conn ->
    let* sum = Exec.add 1 2 conn in
    let* product = Exec.mul 3 4 conn in
    Ok { sum; product }
  in

  match program with
  | Error e -> failwith (sprintf "Got error! %s" (Caqti_error.show e))
  | Ok { sum; product } ->
      print_endline "Program ran successfully!";
      printf "DB said: 1+2=%d and 3*4=%d\n" sum product

let () = Eio_main.run @@ fun env -> main ~env
```

It's a bit more verbose (I don't mind verbose if it's clearer), but I just wanted to make sure that's what you wanted since I wasn't sure you did see the mentionned commit as I re-read your comment.

Reply:
Ah yes, I can see you also mentioned "fully expanded version" in the README.  My main worry is promoting a style for `with_conn` in the README which one would not use in real software.  Maybe we should elaborate the comment above the README version, say something like:?

> Firstly, let's define a variant of `with_conn` which is fused with `Eio_main.run` for convenience when using the toplevel.

And maybe even name it `run_with_conn` to further set it apart from the `Repo` version?

Reply:
Got it. I completely removed eio from lib in my latest commit.